### PR TITLE
Add support  for `Filter_Condition` in `any`, `all`, `find`, `partition` and `index_of`

### LINKED ISSUE #6628: Support `Filter_Condition` in `any`, `all` and `find`
Our `filter` method accepts an Enso predicate OR a `Filter_Condition`. The latter is (supposed to be) easier constructible in the GUI and also possible to translate into SQL for the Table backend.

Apart from the `filter` method, there are also similar methods which take a predicate - `any`, `all` and `find`. I think we should extend them to also accept a `Filter_Condition` - for consistency with `filter`. It should make these methods easier to use in the GUI too.

The change should be uniformly applied to all of our sequence types: 
- [x] `Vector`, 
- [x] `List`, 
- [x] `Array`, 
- [x] `Range`, 
- [x] `Date_Range`.

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

Closes #6628 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso [hunk 1]
```diff
@@ -530,15 +540,20 @@ type Array
...
        Arguments:
-       - predicate: A function that takes an element and returns a boolean
-         value.
+       - condition: A `Filter_Condition` or a function that takes an element and
+         returns a boolean value.
+
+       > Example
+         Splitting an array into elements that start with a prefix.
+
+             ["a", "b", "ax", "bx"].to_array.partition (Filter_Condition.Starts_With "a") == (Pair ["a", "ax"].to_array ["b", "bx"].to_array)
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso:549]:
```suggestion
             ["a", "b", "ax", "bx"].to_array.partition (Filter_Condition.Starts_With "a") == (Pair ["a", "ax"]["b", "bx"])
```

These `to_array` on the right should not be there. They are 'technically correct', because `["a", "ax"].to_array == ["a", "ax"]`, but `Array.partition` returns _vectors_ so IMO they are misleading.

Nevertheless they have to be there to satisfy the `VectorArraySignatureTest`. Finding some better heuristic for the test seems to be out of scope of this PR.

**Comment by radeusgd** [comment]
image

Some error cases may still need better handling:
image

**Comment by radeusgd** [comment]
image
image
image

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso [hunk 2]
```diff
@@ -530,15 +540,20 @@ type Array
        in the input array.
 
        Arguments:
-       - predicate: A function that takes an element and returns a boolean
-         value.
+       - condition: A `Filter_Condition` or a function that takes an element and
+         returns a boolean value.
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso]: Maybe simplify "a function that takes an element and returns a boolean value" to "predicate function" everywhere.

**Review by GregoryTravis — APPROVED** [review]

**Comment by radeusgd** [comment]
New changes ensure that Range and Date Range only get relevant widgets:
image
image
image


And provide better error messages:
image
image
image
image
image

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso (hunk 2 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso]: I was wondering if `predicate` is a well known term for non-programmers?

Since I'm not a native speaker - I have no idea. What do you think? 

From my experience of the Polish counterpart `predykat` - it is something that shows up rather only in computer science, mathematical logic or maybe philosophy literature. I would not assume someone with e.g. purely financial background must be familiar with that term.

**Comment by radeusgd** [comment]
Partition added to `List` for consistency:

image

> Note: We may want to improve how a `List` is visualized inside of a Table vis. Not high priority though since we rarely use lists.

image
image
image

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso (hunk 2 above)
Comment (jdunkerley) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso]: How about?
```
- condition: A `Filter_Condition` or a predicate function to test each element against returning a `Boolean`.
```

**Review by jdunkerley — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso (hunk 2 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso]: I'm happy to rephrase it, but I don't like "test each element against returning a `Boolean`."

I like `test each element` - that part sounds OK.

But `test ... against returning` - to me it sounds like we are testing collection elements to see if each _element is itself a function that may return a boolean_.

(Test something against doing X sounds to me like test if that thing does X - in our case returns boolean).

So IMO that phrasing is also confusing.

Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso]: IMO we should either have `function returning a boolean` or a `predicate` - having both is a pleonasm. (Although I appreciate the SEO :))

Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso]: Settling for:
```
       - condition: A `Filter_Condition` or a predicate function to test each
         element.
```
